src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,174 @@
+            Strategy = options.Strategy
+            LevelOfConcurrency = options.LevelOfConcurrency
+            Count = this.Store.Count
+            MostRecentlyAccesssed = this.Store.Values |> Seq.maxBy _.LastAccessed |> _.LastAccessed
+            LeastRecentlyAccesssed = this.Store.Values |> Seq.minBy _.LastAccessed |> _.LastAccessed
+            MostFrequentlyAccessed = this.Store.Values |> Seq.maxBy _.AccessCount |> _.AccessCount
+            LeastFrequentlyAccessed = this.Store.Values |> Seq.minBy _.AccessCount |> _.AccessCount
+        |}
+
+
+    member private _.GetEvictCount() =
+        if this.Store.Count >= options.MaximumCapacity then
+            (this.Store.Count - options.MaximumCapacity) + (options.MaximumCapacity * options.PercentageToEvict / 100)
+        else
+            0
+
+    // TODO: All of these are proofs of concept, a very naive implementation of eviction strategies, it will always walk the dictionary to find the items to evict, this is not efficient.
+    member private _.TryGetItemsToEvict () =
+        this.Store |>
+            match options.Strategy with
+            | CachingStrategy.LRU ->
+                ConcurrentDictionary.sortBy _.Value.LastAccessed
+            | CachingStrategy.LFU ->
+                ConcurrentDictionary.sortBy _.Value.AccessCount
+        |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+
+    member private _.TryEvictItems () =
+        if this.GetEvictCount() > 0 then
+            for key in this.TryGetItemsToEvict () do
+                match this.Store.TryRemove(key) with
+                | true, _ -> eviction.Trigger(key)
+                | _ -> () // TODO: We probably want to count eviction misses as well?
+
+    // TODO: Shall this be a safer task, wrapping everything in try .. with, so it's not crashing silently?
+    member private this.TryEvictTask () =
+        backgroundTask {
+            while not cts.Token.IsCancellationRequested do
+                    let evictionCount = this.GetEvictCount()
+                    if evictionCount > 0 then
+                        this.TryEvictItems ()
+                    let utilization = (this.Store.Count / options.MaximumCapacity)
+                    // So, based on utilization this will scale the delay between 0 and 1 seconds.
+                    // Worst case scenario would be when 1 second delay happens, then cache will grow rapidly, and beyond the maximum capacity.
+                    // In this case underlying dictionary will resize, AND we will have to evict items, which will likely be slow.
+                    // In this case, cache stats should be used to adjust MaximumCapacity and PercentageToEvict.
+                    let delay = 1000 - (1000 * utilization)
+                    if delay > 0 then
+                        do! Task.Delay(delay)
+
+        }
+
+    // TODO: Explore an eviction shortcut, some sort of list of keys to evict first, based on the strategy.
```

Comment:
This would be quite easy to do - maintain some structure (like linked list), and move pointers to the either end of it based on strategy (most accessed to the end, for example), and chop beginning.

The question is if it's worth to allocate, maintain ans synchronise it.

Reply:
I would say it's worth it, especially if we want to put more items in the cache. Since eviction is super cheap then it can be done on each access. Also we then wouldn't need the `CachedEntity` type. At least for the eviction strategies based on recency. 

I guess the only downside is that we'd have to maintain the lock to ensure thread safety. Probably wouldn't need a ConcurrentDictionary then also.

Anyway, this can always be done as a future improvement. Ideally if we have some benchmarks for it.

Reply:
Yeah, I'm not a huge fan of dealing with concurrency myself tbh, when CD exists and is done well. I'll see what can be done with backlog.